lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -0,0 +1,457 @@
+
+
+def _is_installed_collection_dir(dir_path):
+    b_dir_path = to_bytes(dir_path, errors='surrogate_or_strict')
+    return os.path.isfile(os.path.join(b_dir_path, _MANIFEST_JSON))
+
+
+def _is_collection_dir(dir_path):
+    return (
+        _is_installed_collection_dir(dir_path) or
+        _is_collection_src_dir(dir_path)
+    )
+
+
+def _find_collections_in_subdirs(dir_path):
+    b_dir_path = to_bytes(dir_path, errors='surrogate_or_strict')
+    galaxy_yml_glob_pattern = os.path.join(
+        b_dir_path,
+        # b'*',  # namespace is supposed to be top-level per spec
+        b'*',  # collection name
+        _GALAXY_YAML,
+    )
+    return (
+        os.path.dirname(galaxy_yml)
+        for galaxy_yml in iglob(galaxy_yml_glob_pattern)
+    )
+
+
+def _is_collection_namespace_dir(tested_str):
+    return any(_find_collections_in_subdirs(tested_str))
+
+
+def _is_file_path(tested_str):
+    return os.path.isfile(to_bytes(tested_str, errors='surrogate_or_strict'))
+
+
+def _is_http_url(tested_str):
+    return urlparse(tested_str).scheme.lower() in {'http', 'https'}
+
+
+def _is_git_url(tested_str):
+    return tested_str.startswith(('git+', 'git@'))
+
+
+def _is_concrete_artifact_pointer(tested_str):
+    return any(
+        predicate(tested_str)
+        for predicate in (
+            # NOTE: Maintain the checks to be sorted from light to heavy:
+            _is_git_url,
+            _is_http_url,
+            _is_file_path,
+            _is_collection_dir,
+            _is_collection_namespace_dir,
+        )
+    )
+
+
+def _is_fqcn(tested_str):
+    # FIXME: port this to AnsibleCollectionRef.is_valid_collection_name
+    if tested_str.count('.') != 1:
+        return False
+
+    return all(
+        # FIXME: keywords and identifiers are different in differnt Pythons
+        not iskeyword(ns_or_name) and _is_py_id(ns_or_name)
+        for ns_or_name in tested_str.split('.')
+    )
+
+
+class _ComputedReqKindsMixin:
+
+    @classmethod
+    def from_dir_path_as_unknown(  # type: ignore
+            cls,  # type: Type[Collection]
+            dir_path,  # type: bytes
+            art_mgr,  # type: ConcreteArtifactsManager
+    ):  # type: (...)  -> Collection
+        """Make collection from an unspecified dir type.
+
+        This alternative constructor attempts to grab metadata from the
+        given path if it's a directory. If there's no metadata, it
+        falls back to guessing the FQCN based on the directory path and
+        sets the version to "*".
+
+        It raises a ValueError immediatelly if the input is not an
+        existing directory path.
+        """
+        if not os.path.isdir(dir_path):
+            raise ValueError(
+                "The collection directory '{path!s}' doesn't exist".
+                format(path=to_native(dir_path)),
+            )
+
+        try:
+            return cls.from_dir_path(dir_path, art_mgr)
+        except ValueError:
```

Comment:
`from_dir_path` tries to get the FQCN from the metadata and can cause an IOError if the metadata file doesn't exist:
```
Traceback (most recent call last):
  File "/root/ansible/bin/ansible-galaxy", line 135, in <module>
    exit_code = cli.run()
  File "/root/ansible/lib/ansible/cli/galaxy.py", line 536, in run
    context.CLIARGS['func']()
  File "/root/ansible/lib/ansible/cli/galaxy.py", line 1299, in execute_list
    self.execute_list_collection()
  File "/root/ansible/lib/ansible/cli/galaxy.py", line 68, in method_wrapper
    return wrapped_method(*args, **kwargs)
  File "/root/ansible/lib/ansible/cli/galaxy.py", line 1423, in execute_list_collection
    collection_path, artifacts_manager,
  File "/root/ansible/lib/ansible/galaxy/collection/__init__.py", line 993, in find_existing_collections
    artifacts_manager,
  File "/root/ansible/lib/ansible/galaxy/dependency_resolution/dataclasses.py", line 162, in from_dir_path_as_unknown
    return cls.from_dir_path(dir_path, art_mgr)
  File "/root/ansible/lib/ansible/galaxy/dependency_resolution/dataclasses.py", line 187, in from_dir_path
    req_name = art_mgr.get_direct_collection_fqcn(tmp_inst_req)
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 206, in get_direct_collection_fqcn
    self._get_direct_collection_namespace(collection),  # type: ignore
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 190, in _get_direct_collection_namespace
    return self.get_direct_collection_meta(collection)['namespace']  # type: ignore
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 234, in get_direct_collection_meta
    collection_meta = _get_meta_from_dir(b_artifact_path)
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 477, in _get_meta_from_dir
    return _get_meta_from_installed_dir(b_path)
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 519, in _get_meta_from_installed_dir
    with open(b_manifest_json_path, 'rb') as manifest_fd:
IOError: [Errno 2] No such file or directory: '/tmp/tmp.pOHC9vpM6d/local/ansible_collections/ansible_test/development/MANIFEST.json'
```